airbyte-integrations/connectors/source-pinterest/source_pinterest/reports/reports.py

```diff
@@ -0,0 +1,171 @@
+        """Return the body of the API request in JSON format."""
+        return self._construct_request_body(stream_slice["start_date"], stream_slice["end_date"], self.granularity, get_analytics_columns())
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        """Return the request parameters."""
+        return {}
+
+    def backoff_max_time(func):
+        def wrapped(self, *args, **kwargs):
+            return backoff.on_exception(backoff.constant, RetryableException, max_time=self.report_wait_timeout * 60, interval=10)(func)(
+                self, *args, **kwargs
+            )
+
+        return wrapped
+
+    def backoff_max_tries(func):
+        def wrapped(self, *args, **kwargs):
+            return backoff.on_exception(backoff.expo, ReportGenerationFailure, max_tries=self.report_generation_maximum_retries)(func)(
+                self, *args, **kwargs
+            )
+
+        return wrapped
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        """Read the records from the stream."""
+        report_infos = self._init_reports(super().read_records(sync_mode, cursor_field, stream_slice, stream_state))
+        self._try_read_records(report_infos, stream_slice)
+
+        for report_info in report_infos:
+            metrics = report_info.metrics
+            for campaign_id, records in metrics.items():
+                self.logger.info(f"Reports for campaign id: {campaign_id}:")
+                yield from records
+
+    @backoff_max_time
+    def _try_read_records(self, report_infos, stream_slice):
+        """Try to read the records and raise appropriate exceptions in case of failure or in-progress status."""
+        incomplete_report_infos = self._incomplete_report_infos(report_infos)
+        for report_info in incomplete_report_infos:
+            report_status, report_url = self._verify_report_status(report_info, stream_slice)
+            report_info.report_status = report_status
+            if report_status in {ReportStatus.DOES_NOT_EXIST, ReportStatus.EXPIRED, ReportStatus.FAILED, ReportStatus.CANCELLED}:
+                message = "Report generation failed."
+                raise ReportGenerationFailure(message)
+            elif report_status == ReportStatus.FINISHED:
+                try:
+                    report_info.metrics = self._fetch_report_data(report_url)
+                except requests.HTTPError as error:
+                    raise ReportGenerationFailure(error)
+
+        pending_report_status = [report_info for report_info in report_infos if report_info.report_status != ReportStatus.FINISHED]
+
+        if len(pending_report_status) > 0:
```

Comment:
this should be changed as well if we're going to implement a queue of reports

Reply:
https://github.com/airbytehq/airbyte/pull/28672#discussion_r1274853525